devito/mpi/distributed.py

```diff
@@ -421,6 +440,226 @@ def _rebuild(self, shape=None, dimensions=None, comm=None):
+        The subdomain to be decomposed.
+
+    Notes
+    -----
+    This class is used internally by Devito for distributing
+    Functions defined on SubDomains. It stores reference to the
+    parent Distributor from which it is created.
+    """
+
+    def __init__(self, subdomain):
+        # Does not keep reference to the SubDomain since SubDomain will point to
+        # this SubDistributor and Distributor does not point to the Grid
+
+        super().__init__(subdomain.shape, subdomain.dimensions)
+
+        self._dimension_map = frozendict({pd: sd for pd, sd
+                                          in zip(subdomain.grid.dimensions,
+                                                 subdomain.dimensions)})
+
+        self._parent = subdomain.grid.distributor
+
+        self._comm = self.parent.comm
+        self._topology = self.parent.topology
+
+        self.__decomposition_setup__()
+
+    def __decomposition_setup__(self):
+        """
+        Set up the decomposition, aligned with that of the parent Distributor.
+        """
+        def interval_bounds(interval):
+            """Extract SubDimension Interval bounds."""
+            if interval.is_empty:
+                # SubDimension has no indices. Min and max are NaN.
+                return np.NaN, np.NaN
+            elif interval.is_Interval:
+                # Interval containing two or more indices. Min and max are ends.
+                return interval.start, interval.end
+            else:
+                # Interval where start == end defaults to FiniteSet.
+                # Repeat this value for min and max
+                return interval.args[0], interval.args[0]
+
+        decompositions = []
+        for dec, i in zip(self.parent._decomposition, self._sd_interval):
+            if i is None:
+                decompositions.append(dec)
+            else:
+                start, end = interval_bounds(i)
+                decompositions.append([d[np.logical_and(d >= start, d <= end)]
+                                       for d in dec])
+
+        self._decomposition = [Decomposition(d, c)
+                               for d, c in zip(decompositions, self.mycoords)]
+
+    @property
+    def parent(self):
+        """The parent distributor of this SubDistributor."""
+        return self._parent
+
+    @property
+    def par_shape(self):
+        """Shape of the parent decomposition."""
+        return self.parent.glb_shape
+
+    @property
+    def par_dimensions(self):
+        """Dimensions of the parent decomposition."""
+        return self.parent.dimensions
+
+    @property
+    def par_slices(self):
+        """
+        The global indices owned by the calling MPI rank, as a mapper from
+        Dimensions to slices. Shortcut for `parent.glb_slices`.
+        """
+        return self.parent.glb_slices
+
+    @property
+    def dimension_map(self):
+        return self._dimension_map
+
+    @cached_property
+    def _d_interval(self):
```

Comment:
btw `_domain_interval` would be clearer, I wouldn't have guessed it w/o a `_sd_interval`